basic_pipeline/6_OrderingBuffer.md

```diff
@@ -0,0 +1,166 @@
++ `seq` is a sequence of 's', 'e', 'q' characters (we need to adjust our regex description to match the header of the packet)
++ `\:` stands for the ':' character (we also need to escape that character since it is meaningful in the regex's syntax)
++ `(?<seq_id>\d+)` allows us to define a named capture - later one, once we use the `Regex.named_captures/2`, we will retrieve the map with 'seq_id' key and the corresponding value equal to the string described by the `\d+` 'partial' regex (which means - one or more occurrences of a decimal). Generally speaking, a named capture can be specified with the following structure: `(?<key>regex)` where `regex` is a regex description.
++ `\]` is the escaped closing square bracket character
++ `(?<data>.*)` is a named capture description that allows us to get the value of a `.*` regex (any character no or any number of times) under a `data` key.
++ `$` stands for the end of the input
+
+The result of `Regex.named_captures/2` applied to that regex description and the exemplary packet should be following:
+```Elixir
+{"seq_id"=>7, "data"=>"[frameid:2][timestamp:3]data"}
+```
+
+Once we unzip the header of the packet in the `handle_process/4` callback, we can put the incoming packet in the `ordered_packets` list and sort that list. Due to the fact, that elements of this list are tuples, whose first element is a sequence id (a value that is unique), the list will be sorted based on the sequence id.
+We also get the sequence id of the first element in the updated `ordered_packets` list.
+
+
+Here comes the rest of the `handle_process/4` definition:
+```Elixir
+# FILE: lib/elements/OrderingBuffer.ex
+
+defmodule Basic.Elements.OrderingBuffer do
+  ...
+  def handle_process(:input, buffer, _context, state) do
+  ...
+    if state.last_processed_seq_id + 1 == last_seq_id do
+      {reversed_ready_packets_sequence, ordered_packets} = get_ready_packets_sequence(ordered_packets, [])
+      [{last_processed_seq_id, _} | _] = reversed_ready_packets_sequence
+
+      state = %{
+        state
+        | ordered_packets: ordered_packets,
+          last_processed_seq_id: last_processed_seq_id
+      }
+      buffers = Enum.reverse(reversed_ready_packets_sequence) |> Enum.map(fn {_seq_id, data} -> data end)
+
+      {{:ok, buffer: {:output, buffers}}, state}
+    else
+      {{:ok, redemand: :output}, state}
+    end
+  end
+  ...
+end
+```
+
+We need to distinguish between two situations: the currently processed packet can have a sequence id which is subsequent to the sequence id of the last processed packet or there might be some packets not yet delivered to us, with sequence ids in between the last processed sequence id and the sequence id of a currently processed packet. In case the second situation occurs, we should store the packet and wait for the next packets to arrive.
```

Comment:
last processed packet collides with handle_process, packets that end up in the list of ordered packets have also been processed in terms of handle_process right? Maybe it should be the last sent packet or sth like this.

Reply:
I have renamed it both in the code and in the text to last sent packet / last sent packet sequence id

Reply:
done